src/generators/gradient_based/loss.jl

```diff
@@ -39,9 +37,15 @@ function ∇(
     M::Models.AbstractDifferentiableModel,
     ce::AbstractCounterfactualExplanation,
 )
-    ℓ = 0
-    if (ce.convergence[:converge_when] == :invalidation_rate)
-        ℓ = hinge_loss(ce)
-    end
-    return ∂ℓ(generator, M, ce) + ∂h(generator, ce) .+ ℓ
+    return ∂ℓ(generator, M, ce) + ∂h(generator, ce) .+ hinge_loss(ce.convergence, ce)
```

Comment:
Actually I just stumbled across this. That can't be right? Now it looks like every gradient-based generator also get's hinge loss?

Reply:
Oh I see, you overloaded that method to return 0 by default. Hmmm let me have a quick go at this before we merge please.

Reply:
Also doesn't look like Probe is now actually using these loss functions (correct me if I'm missing something)

```julia
function ProbeGenerator(;
    λ::AbstractFloat=0.1,
    loss::Symbol=:logitbinarycrossentropy,
    penalty::Penalty=Objectives.distance_l1,
    kwargs...,
)
    @assert haskey(losses_catalogue, loss) "Loss function not found in catalogue."
    user_loss = Objectives.losses_catalogue[loss]
    return GradientBasedGenerator(; loss=user_loss, penalty=penalty, λ=λ, kwargs...)
end
```

Reply:
Another thing that doesn't seem to check out above is that hinge loss isn't differentiated. We should improve tests here to catch such things.

Reply:
> Also doesn't look like Probe is now actually using these loss functions (correct me if I'm missing something)

This is correct, the user has to specify `InvalidationRateConvergence` as the convergence type to use Probe in the intended way. It's the same way it was before: though `hinge_loss()` was defined in the same folder as `ProbeGenerator`, Probe actually called the `hinge_loss()` method only if the user defined `converge_when = :invalidation_rate` inside the core struct (see the deleted line 43 in `src/generators/gradient_based/loss.jl`). The reason I'm dispatching on `hinge_loss()` based on the convergence type rather than based on whether the user defined Probe as the generator is that I want the generators to remain composable: this way, you can also use `hinge_loss()` and `invalidation_rate` with other generators if you want.